crates/edr_evm_napi/src/tracer/js_tracer.rs

```diff
@@ -0,0 +1,490 @@
+use std::{
+    fmt::Debug,
+    sync::mpsc::{channel, Sender},
+};
+
+use edr_eth::Bytes;
+use edr_evm::{
+    return_revert, trace::BeforeMessage, Bytecode, CallInputs, EVMData, Gas, InstructionResult,
+    SuccessOrHalt,
+};
+use napi::{Env, JsBufferValue, JsFunction, JsUndefined, NapiRaw, Status};
+use napi_derive::napi;
+
+use crate::{
+    sync::{await_promise, handle_error},
+    threadsafe_function::{ThreadSafeCallContext, ThreadsafeFunction, ThreadsafeFunctionCallMode},
+    trace::TracingMessage,
+    transaction::result::{ExceptionalHalt, ExecutionResult},
+};
+
+#[napi(object)]
+pub struct TracingCallbacks {
+    #[napi(ts_type = "(message: TracingMessage, next: any) => Promise<TracingMessage>")]
+    pub before_call: JsFunction,
+    #[napi(ts_type = "(result: ExecutionResult, next: any) => Promise<ExecutionResult>")]
+    pub after_call: JsFunction,
+}
+
+struct BeforeCallEvent {
+    data: BeforeMessage,
+    sender: Sender<napi::Result<BeforeMessage>>,
+}
+
+pub struct AfterCallEvent {
+    pub result: edr_evm::ExecutionResult,
+    pub sender: Sender<napi::Result<edr_evm::ExecutionResult>>,
+}
+
+#[derive(Clone)]
+pub struct JsTracer {
+    before_call_fn: ThreadsafeFunction<BeforeCallEvent>,
+    after_call_fn: ThreadsafeFunction<AfterCallEvent>,
+    pending_before: Option<BeforeMessage>,
+}
+
+impl JsTracer {
+    /// Constructs an instance from the provided callbacks.
+    pub fn new(env: &Env, callbacks: TracingCallbacks) -> napi::Result<Self> {
+        let before_call_fn = ThreadsafeFunction::create(
+            env.raw(),
+            unsafe { callbacks.before_call.raw() },
+            0,
+            |mut ctx: ThreadSafeCallContext<BeforeCallEvent>| {
```

Comment:
I find it difficult to parse what's going on in these function bodies. 😅 Could you add some high level comments both in the before and after calls explaining the flow please?

Reply:
Hopefully, addressed in https://github.com/NomicFoundation/hardhat/pull/4487/commits/4f7dd4ae886923393968a8b060ab71b2a9607b1c